### PR TITLE
DPO -> FSDP2

### LINKED ISSUE #1404: DPO 🤝 FSDP2
## The what and the why

We have started migrating our recipes onto FSDP2 APIs, as they offer better composability with torch compile, quantization APIs, frozen/unfrozen weights, and more. See the original PyTorch RFC for more details: https://github.com/pytorch/pytorch/issues/114299.

Thus far we migrated our full finetune recipe in #1287, and we already have a LoRA recipe on FSDP2 that just needs to be moved out of our dev/ folder. 

This issue is to track migration of our DPO recipe onto FSDP2, which is the last remaining piece to get us fully onto FSDP2. Once this happens, there are a bunch of FSDP1 APIs in our repo that we can clean up. We can also clean up stuff like [this](https://github.com/pytorch/torchtune/blob/3d7ca386902ac171ef11e786fb60994b98bc495c/torchtune/models/llama3_1/_position_embeddings.py#L121-L124), which screws with compile.

## The how 

This can be done by following the `lora_finetune_fsdp2` recipe. The most important changes will be in the [_setup_model method](https://github.com/pytorch/torchtune/blob/3d7ca386902ac171ef11e786fb60994b98bc495c/recipes/dev/lora_finetune_fsdp2.py#L267-L372). As in that recipe, we need to:

1) instantiate the model on meta device ([ref](https://github.com/pytorch/torchtune/blob/3d7ca386902ac171ef11e786fb60994b98bc495c/recipes/dev/lora_finetune_fsdp2.py#L296-L297))
2) set trainable params and define the activation checkpointing strategy ([ref](https://github.com/pytorch/torchtune/blob/3d7ca386902ac171ef11e786fb60994b98bc495c/recipes/dev/lora_finetune_fsdp2.py#L300-L305))
3) iterate over submodules from lowest to highest, sharding as we go. #1287 introduced the [shard_model](https://github.com/pytorch/torchtune/blob/3d7ca386902ac171ef11e786fb60994b98bc495c/torchtune/utils/_distributed.py#L555) utility, which is applied for full finetune [here](https://github.com/pytorch/torchtune/blob/3d7ca386902ac171ef11e786fb60994b98bc495c/recipes/full_finetune_distributed.py#L369-L397). The only difference for the DPO recipe is that we will need to add one extra condition to shard the LoRA A and B params (as done [here](https://github.com/pytorch/torchtune/blob/3d7ca386902ac171ef11e786fb60994b98bc495c/recipes/dev/lora_finetune_fsdp2.py#L315-L316) in the LoRA FSDP2 recipe)
4) load LoRA weights if provided ([ref](https://github.com/pytorch/torchtune/blob/3d7ca386902ac171ef11e786fb60994b98bc495c/recipes/dev/lora_finetune_fsdp2.py#L327-L332))
5) perform custom initialization of any parameters that need it (RoPE buffers and optionally LoRA weights) ([ref](https://github.com/pytorch/torchtune/blob/3d7ca386902ac171ef11e786fb60994b98bc495c/recipes/dev/lora_finetune_fsdp2.py#L334-L345))
6) load the base model weights ([ref](https://github.com/pytorch/torchtune/blob/3d7ca386902ac171ef11e786fb60994b98bc495c/recipes/dev/lora_finetune_fsdp2.py#L347-L349))

**Testing**: This is one of our only recipes without CI 😢. We should run the recipe before and after the changes on a couple different models to ensure (a) it works, (b) loss curves are unchanged, and (c) memory and perf are comparable.

### PR DESCRIPTION
**Pull request by ebsmothers**
Fixes #1404 

Last recipe to move to FSDP2. Also a little cleanup of docstrings and two fixes in our distributed full finetune recipe

## Test plan:

This one only has one config 😅 : Llama2-7B. Also ran with `fsdp_cpu_offload=True` and everything works fine there.

### Baseline

```
tune run --nnodes 1 --nproc_per_node 2 lora_dpo_distributed --config llama2/7B_lora_dpo \
metric_logger=torchtune.training.metric_logging.WandBLogger metric_logger.project=lora-dpo-fsdp2 \
metric_logger.name=baseline max_steps_per_epoch=100 gradient_accumulation_steps=2 epochs=1 \
log_peak_memory_stats=True
```

### This PR

```
tune run --nnodes 1 --nproc_per_node 2 lora_dpo_distributed --config llama2/7B_lora_dpo \
metric_logger=torchtune.training.metric_logging.WandBLogger metric_logger.project=lora-dpo-fsdp2 \
metric_logger.name=FSDP2 max_steps_per_epoch=100 gradient_accumulation_steps=2 epochs=1 \
log_peak_memory_stats=True
```

### Results
Screenshot 2024-09-10 at 3 49 06 PM

### CODE REVIEW

**Comment by felipemello1** [comment]
just for sanity check, would it be worth running with cpu_offload = True?

**Comment by ebsmothers** [comment]
> just for sanity check, would it be worth running with cpu_offload = True?

Oh yeah I ran and it succeeded, just didn't include the results here (also I didn't log them). But I can re-run and include when I get a chance

**Comment by felipemello1** [comment]
no need to paste it here! if it ran, thats fine! just wanted to make sure. Thank you!

**Review by joecummings — APPROVED** [review]
